### PR TITLE
catboost: init at 0.24.4

### LINKED ISSUE #74395: Catboost
Would be a very useful package to have for ML Enthusiasts

**Project description**
Machine Learning Catboost Library

**Metadata**
CatBoost is a high-performance open source library for gradient boosting on decision trees

* homepage URL: https://catboost.ai/
* source URL: https://github.com/catboost/catboost
* license: Apache License 2.0 https://github.com/catboost/catboost/blob/master/LICENSE
* platforms: Linux
* installation instructions: https://catboost.ai/docs/installation/python-installation-method-build-from-source-linux-macos.html#python-installation-method-build-from-source-linux-macos

@jonringer Please let me know if I should stop tagging you in random issues and it bothers you (sorry, i'll stop), but I thought this might be of interest to you since you are very involved in tensorflow.

### PR DESCRIPTION
**Pull request by PlushBeaver**
###### Motivation for this change

* Closes #74395
* Personal interest in decision trees with categorical features.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
  - Tested examples from CatBoost tutorial.
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -7089,6 +7089,12 @@
     githubId = 13000278;
     name = "Maksim Bronsky";
   };
+  PlushBeaver = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:7092]: Please put this into a separate commit with the message ``maintainers: add PlushBeaver``.

File: pkgs/development/python-modules/catboost/default.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+buildPythonPackage {
+  pname = "catboost";
+  version = "0.24.4";
+
+  disabled = pythonOlder "3.4";
+
+  src = fetchFromGitHub {
+    owner = "catboost";
+    repo = "catboost";
+    rev = "2cb3eda7dd228e851d9817eb018b9d7521ac342d";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:14]:
```suggestion
    rev = "v${version}";
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    sha256 = "sha256-pzmwEiKziB4ldnKgeCsP2HdnisX8sOkLssAzNfcSEx8=";
+  };
+
+  # Passed to make in setup.py.
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:18]:
```suggestion
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  nativeBuildInputs = [ clang_7 python2 ];
+
+  propagatedBuildInputs = [ graphviz matplotlib numpy pandas scipy plotly six ]
+    ++ lib.optional withCuda [ cudatoolkit ];
+
+  # First apply patches, then go to source root; sourceRoot applies too early.
+  patches = [ ./nix-support.patch ./format.patch ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:25]:
```suggestion
  patches = [
    ./nix-support.patch
    ./format.patch
  ];
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  postPatch = "cd catboost/python-package";
+
+  setupPyBuildFlags = [ "--with-ymake=no" ];
+  CUDA_ROOT = lib.optional withCuda cudatoolkit;
+
+  # Tests use custom "ya" tool, not yet supported.
+  dontUseSetuptoolsCheck = true;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:32]:
```suggestion
  dontUseSetuptoolsCheck = true;
  pythonImportsCheck = [ "catboost" ];
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  # Would use setup.py --parallel=..., but setupPyBuildFlags are quoted.
+  MAKEFLAGS = "-j$(NIX_BUILD_CORES)";
+  enableParallelBuilding = true;
+
+  meta = {
+    description =
+      "High-performance library for gradient boosting on decision trees.";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:40]:
```suggestion
    description = "High-performance library for gradient boosting on decision trees";
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  CUDA_ROOT = lib.optional withCuda cudatoolkit;
+
+  # Tests use custom "ya" tool, not yet supported.
+  dontUseSetuptoolsCheck = true;
+
+  # Would use setup.py --parallel=..., but setupPyBuildFlags are quoted.
+  MAKEFLAGS = "-j$(NIX_BUILD_CORES)";
+  enableParallelBuilding = true;
+
+  meta = {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:38]:
```suggestion
  meta = with lib; {
```

File: pkgs/development/python-modules/catboost/default.nix [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      "High-performance library for gradient boosting on decision trees.";
+    longDescription = ''
+      A fast, scalable, high performance Gradient Boosting on Decision Trees
+      library, used for ranking, classification, regression and other machine
+      learning tasks for Python, R, Java, C++. Supports computation on CPU and GPU.
+    '';
+    license = lib.licenses.asl20;
+    platforms = [ "x86_64-linux" ];
+    homepage = "https://catboost.ai";
+    maintainers = [ lib.maintainers.PlushBeaver ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:49]:
```suggestion
    license = licenses.asl20;
    platforms = [ "x86_64-linux" ];
    homepage = "https://catboost.ai";
    maintainers = with maintainers; [ PlushBeaver ];
```

File: pkgs/development/python-modules/catboost/format.patch [hunk 9]
```diff
@@ -0,0 +1,15 @@
+Don't use variable format string, which triggers -Wformat-security.
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/format.patch:1]: Can you create a upstream PR and use fetchpatch?

File: pkgs/development/python-modules/catboost/nix-support.patch [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+--- a/make/python2.CLANG7-LINUX-X86_64.makefile
++++ b/make/python2.CLANG7-LINUX-X86_64.makefile
+@@ -12,8 +12,8 @@ endef
+ _CC_VERSION = $(shell echo '$(_CC_TEST)' | $(CC) -E -P -)
+ $(info _CC_VERSION = '$(_CC_VERSION)')
+ 
+-ifneq '$(_CC_VERSION)' '7 0'
+-    $(error clang 7.0 is required)
++ifneq '$(_CC_VERSION)' '7 1'
++    $(error clang 7.1 is required)
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/nix-support.patch:32]: Why bump the clang version?

File: pkgs/development/python-modules/catboost/nix-support.patch [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+         logging.info('Buildling {} with gnu make'.format(catboost_ext))
+-        makefile = 'python{}.{}CLANG50-LINUX-X86_64.makefile'.format(python_version()[0], 'CUDA.' if self.with_cuda else '')
++        makefile = 'python{}.{}CLANG7-LINUX-X86_64.makefile'.format(python_version()[0], 'CUDA.' if self.with_cuda else '')
+         make_cmd = [
+             'make', '-f', '../../make/' + makefile,
+-            'CC=clang-5.0',
+-            'CXX=clang++-5.0',
++            'CC=clang',
++            'CXX=clang++',
++            'PYTHON=python2',
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/nix-support.patch:17]: What about python3?

File: pkgs/development/python-modules/catboost/default.nix [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  postPatch = "cd catboost/python-package";
+
+  setupPyBuildFlags = [ "--with-ymake=no" ];
+  CUDA_ROOT = lib.optional withCuda cudatoolkit;
+
+  # Tests use custom "ya" tool, not yet supported.
+  dontUseSetuptoolsCheck = true;
+
+  # Would use setup.py --parallel=..., but setupPyBuildFlags are quoted.
+  MAKEFLAGS = "-j$(NIX_BUILD_CORES)";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:35]:
```suggestion
  MAKEFLAGS = "-j$NIX_BUILD_CORES";
```

I think caused some issue why my machine froze.

File: pkgs/development/python-modules/catboost/format.patch (hunk 9 above)
Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/format.patch:1]: PR: https://github.com/catboost/catboost/pull/1528
I'll look into using `fetchpatch`.

File: pkgs/development/python-modules/catboost/nix-support.patch (hunk 11 above)
Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/nix-support.patch:17]: Python 2 is required for build scripts, they fail with Python 3. CatBoost itself can build for 2 & 3, but pyrsistent-0.17.3 not supported for interpreter python2.7 (pandas dependency).

File: pkgs/development/python-modules/catboost/nix-support.patch (hunk 10 above)
Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/nix-support.patch:32]: We have clang 7.1.0 in 20.03 and 20.09. I think upstream states 7.0, because it's in Ubuntu 20.04.

File: pkgs/development/python-modules/catboost/default.nix (hunk 4 above)
Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/default.nix:25]: Re: all formatting suggestions. I assumed nixfmt would yield a satisfying result. Is there a style guide? I'll follow your suggestions anyway.

File: pkgs/development/python-modules/catboost/default.nix (hunk 12 above)
Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/default.nix]: Your suggestion effetively disables parallel build. Used preBuild to set MAKEFLAGS. On my 8-core machine it speeds up the build without any freezes.

File: pkgs/development/python-modules/catboost/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/catboost/default.nix]: Not really a comprehensive one.

File: pkgs/development/python-modules/catboost/nix-support.patch (hunk 10 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/catboost/nix-support.patch:32]: I would be fine with removing the check but bumping it seems a little to much maintenance work but should work.

File: pkgs/development/python-modules/catboost/default.nix (hunk 12 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/catboost/default.nix]: NIX_BUILD_CORE is a global variable by nix and not a command. Maybe the space is not optional like I thought but the () are definitely to much here. Maybe your nix-build-cores settings is wrong.

File: pkgs/development/python-modules/catboost/nix-support.patch (hunk 10 above)
Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/nix-support.patch]: Checks removed.

File: pkgs/development/python-modules/catboost/default.nix (hunk 12 above)
Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/default.nix]: Modified nix support patch, so that setup.py picks up NIX_BUILD_CORES and enableParallelBuilding just works.

File: pkgs/development/python-modules/catboost/default.nix [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+      formatPatch = fetchpatch {
+        name = "format.patch";
+        url = "https://github.com/catboost/catboost/pull/1528.patch";
+        sha256 = "sha256-fNGucHxsSDFRLk3hFH7rm+zzTdDpY9/QjRs8K+AzVvo=";
+      };
+    in [
+      ./nix-support.patch
+      formatPatch
+    ];
+  postPatch = "cd catboost/python-package";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:35]:
```suggestion
  patches = [
    ./nix-support.patch
    (fetchpatch {
      name = "format.patch";
      url = "https://github.com/catboost/catboost/pull/1528.patch";
      sha256 = "sha256-fNGucHxsSDFRLk3hFH7rm+zzTdDpY9/QjRs8K+AzVvo=";
    };)
  ];

```

File: pkgs/development/python-modules/catboost/default.nix [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:36]:
```suggestion

preBuild = ''
  cd catboost/python-package
  '';

```

File: pkgs/development/python-modules/catboost/default.nix [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  nativeBuildInputs = [ clang_7 python2 ];
+
+  propagatedBuildInputs = [ graphviz matplotlib numpy pandas scipy plotly six ]
+    ++ lib.optional withCuda [ cudatoolkit ];
+
+  patches = [
+    ./nix-support.patch
+    (fetchpatch {
+      name = "format.patch";
+      url = "https://github.com/catboost/catboost/pull/1528.patch";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:28]: Please fetch specific commits and not PRs.

**Comment by roberth** [comment]
@GrahamcOfBorg build pythonPackages.catboost

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108186` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>python37Packages.catboost</li>
<li>python38Packages.catboost</li>
</ul>
</details>

File: pkgs/development/python-modules/catboost/default.nix [hunk 16]
```diff
@@ -0,0 +1,56 @@
+{ buildPythonPackage, fetchFromGitHub, fetchpatch, lib, pythonOlder
+, clang_7, python2
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/catboost/default.nix:2]: I must have been drunk when merging this. We can't use python2 or python3 as inputs in pythonPackages.

Comment (PlushBeaver) [reply pkgs/development/python-modules/catboost/default.nix:2]: Upstream scripts invoked during the build of native modules are python2-only. I don't think that upstream developers are willing to upgrade them because CatBoost still supports Python 2 installation.
